shell/platform/android/io/flutter/view/FlutterMain.java

```diff
@@ -332,10 +333,18 @@ public static boolean isRunningPrecompiledCode() {
         return sIsPrecompiledAsBlobs || sIsPrecompiledAsSharedLibrary;
     }
 
-    public static String findAppBundlePath(Context applicationContext) {
+    @NonNull
+    public static String findAppBundlePath(@NonNull Context applicationContext) {
         String dataDirectory = PathUtils.getDataDirectory(applicationContext);
         File appBundle = new File(dataDirectory, sFlutterAssetsDir);
-        return appBundle.exists() ? appBundle.getPath() : null;
+        if (!appBundle.exists()) {
```

Comment:
I have heard multiple times from Jason that FlutterMain should go away.

@jason-simmons How could we write this code not to use FlutterMain?

Reply:
Revisiting FlutterMain is something I'll do after this PR.  Right now its just about FlutterActivity and FlutterFragment and that work is really just about pushing as much existing behavior from th eActivity to the Fragment as possible.  There is certainly still plenty of rework to be done.

Reply:
It would be great if applications didn't have to initialize the engine up front and do the `startInitialization`/`ensureInitializationComplete` sequence.  However, I don't know if we have a viable way to avoid it.

The main thing that needs to happen in FlutterMain is extraction of resources (specifically the ICU data file) out of the APK and into local disk storage that the engine can then mmap.

The alternatives would be storing icudtl.dat uncompressed in the APK and mapping it directly from the APK (which would increase APK size).  Or decompressing icudtl.dat into a buffer (which would increase the app's memory usage).

We would also need to change some engine global settings to be per-instance and/or lazily initialize them when the first engine instance starts up.

@chinmaygarde

Reply:
Why would postponing the extraction step until later require any other changes?  Yes, the extraction needs to occur before anything else happens, but why can't we simply let the developer decide when to do that?